File: tools/extract_i18n_languages.py
#!/usr/bin/env python3
"""
提取 i18n.js 中的语言文件为独立的 JSON 文件
使用 Python 更安全地解析 JavaScript 对象
"""

import json
import re
import os
from pathlib import Path

# 路径配置
script_dir = Path(__file__).parent
project_root = script_dir.parent
i18n_js_path = project_root / 'frontend/static/js/i18n.js'
locales_dir = project_root / 'frontend/static/js/locales'

# 确保 locales 目录存在
locales_dir.mkdir(parents=True, exist_ok=True)

# 读取 i18n.js 文件
with open(i18n_js_path, 'r', encoding='utf-8') as f:
    content = f.read()

# 找到 i18n 对象的开始和结束位置
i18n_start = content.find('const i18n = {')
if i18n_start == -1:
    print('❌ Could not find i18n object')
    exit(1)

# 找到 LanguageManager 类的位置（这是 i18n 对象的结束标志）
class_start = content.find('class LanguageManager')
if class_start == -1:
    print('❌ Could not find LanguageManager class')
    exit(1)

# 提取 i18n 对象部分
i18n_obj_str = content[i18n_start + len('const i18n = '):class_start].strip()
# 移除末尾的分号
i18n_obj_str = re.sub(r';\s*$', '', i18n_obj_str)

# 定义要提取的语言列表
languages = ['en', 'ru', 'pt', 'es', 'fr', 'ko', 'de', 'ja']

print('Extracting language files...\n')

for lang in languages:
    # 查找语言对象的开始位置
    lang_pattern = re.compile(rf'\s+{re.escape(lang)}:\s*{{')
    match = lang_pattern.search(i18n_obj_str)
    
    if not match:
        print(f'⚠ Language {lang} not found')
        continue
    
    start_idx = match.start()
    
    # 找到第一个 { 的位置
    obj_start = start_idx + len(match.group()) - 1
    for i in range(obj_start, len(i18n_obj_str)):
        if i18n_obj_str[i] == '{':
            obj_start = i
            break
    
    # 查找匹配的结束大括号
    brace_count = 0
    in_string = False
    string_char = None
    escape_next = False
    
    for i in range(obj_start, len(i18n_obj_str)):
        char = i18n_obj_str[i]
        
        if escape_next:
            escape_next = False
            continue
        
        if char == '\\':
            escape_next = True
            continue
        
        if not in_string and char in ('"', "'", '`'):
            in_string = True
            string_char = char
            continue
        
        if in_string and char == string_char:
            in_string = False
            string_char = None
            continue
        
        if not in_string:
            if char == '{':
                brace_count += 1
            elif char == '}':
                brace_count -= 1
                if brace_count == 0:
                    # 找到匹配的结束大括号
                    end_idx = i + 1
                    lang_obj_str = i18n_obj_str[obj_start:end_idx]
                    
                    # 尝试解析为 JSON
                    try:
                        # 先尝试直接解析（如果格式正确）
                        lang_data = json.loads(lang_obj_str)
                        
                        # 保存为 JSON 文件
                        json_path = locales_dir / f'{lang}.json'
                        with open(json_path, 'w', encoding='utf-8') as f:
                            json.dump(lang_data, f, ensure_ascii=False, indent=2)
                        print(f'✓ Extracted {lang}.json')
                    except json.JSONDecodeError:
                        # 如果直接解析失败，尝试使用 JavaScript 解析器
                        # 这里我们使用一个更简单的方法：使用 node 来解析
                        import subprocess
                        import tempfile
                        
                        try:
                            # 创建一个临时 JS 文件
                            with tempfile.NamedTemporaryFile(mode='w', suffix='.js', delete=False) as tmp:
                                tmp.write(f'console.log(JSON.stringify({lang_obj_str}))')
                                tmp_path = tmp.name
                            
                            # 使用 node 执行
                            result = subprocess.run(
                                ['node', tmp_path],
                                capture_output=True,
                                text=True,
                                timeout=5
                            )
                            
                            if result.returncode == 0:
                                lang_data = json.loads(result.stdout.strip())
                                
                                # 保存为 JSON 文件
                                json_path = locales_dir / f'{lang}.json'
                                with open(json_path, 'w', encoding='utf-8') as f:
                                    json.dump(lang_data, f, ensure_ascii=False, indent=2)
                                print(f'✓ Extracted {lang}.json')
                            else:
                                print(f'✗ Failed to extract {lang}: {result.stderr}')
                        except Exception as e:
                            print(f'✗ Failed to extract {lang}: {str(e)}')
                        finally:
                            # 清理临时文件
                            try:
                                os.unlink(tmp_path)
                            except:
                                pass
                    break

print('\n✅ Language extraction completed!')
print(f'📁 Files saved to: {locales_dir}')

File: tools/extract_i18n_languages.js
#!/usr/bin/env node
/**
 * 提取 i18n.js 中的语言文件为独立的 JSON 文件
 * 使用更安全的字符串解析方法
 */

const fs = require('fs');
const path = require('path');

const i18nJsPath = path.join(__dirname, '../frontend/static/js/i18n.js');
const localesDir = path.join(__dirname, '../frontend/static/js/locales');

// 确保 locales 目录存在
if (!fs.existsSync(localesDir)) {
    fs.mkdirSync(localesDir, { recursive: true });
}

// 读取 i18n.js 文件
let content = fs.readFileSync(i18nJsPath, 'utf-8');

// 找到 i18n 对象的开始和结束位置
const i18nStart = content.indexOf('const i18n = {');
if (i18nStart === -1) {
    console.error('Could not find i18n object');
    process.exit(1);
}

// 找到 LanguageManager 类的位置（这是 i18n 对象的结束标志）
const classStart = content.indexOf('class LanguageManager');
if (classStart === -1) {
    console.error('Could not find LanguageManager class');
    process.exit(1);
}

// 提取 i18n 对象部分（不包括最后的分号）
const i18nObjStr = content.substring(i18nStart + 'const i18n = '.length, classStart).trim();
// 移除末尾的分号
const cleanedI18n = i18nObjStr.replace(/;\s*$/, '');

// 定义一个函数来提取单个语言对象
function extractLanguageObject(langKey, content) {
    // 找到语言键的位置
    const langPattern = new RegExp(`\\s+${langKey}:\\s*\\{`);
    const match = content.match(langPattern);
    if (!match) {
        return null;
    }
    
    const startIndex = content.indexOf(match[0]);
    if (startIndex === -1) {
        return null;
    }
    
    // 从开始位置查找匹配的大括号
    let braceCount = 0;
    let inString = false;
    let stringChar = null;
    let escapeNext = false;
    
    for (let i = startIndex + match[0].length; i < content.length; i++) {
        const char = content[i];
        
        if (escapeNext) {
            escapeNext = false;
            continue;
        }
        
        if (char === '\\') {
            escapeNext = true;
            continue;
        }
        
        if (!inString && (char === '"' || char === "'" || char === '`')) {
            inString = true;
            stringChar = char;
            continue;
        }
        
        if (inString && char === stringChar) {
            inString = false;
            stringChar = null;
            continue;
        }
        
        if (!inString) {
            if (char === '{') {
                braceCount++;
            } else if (char === '}') {
                if (braceCount === 0) {
                    // 找到匹配的结束大括号
                    const endIndex = i + 1;
                    const langObjStr = content.substring(startIndex + match[0].length - 1, endIndex);
                    
                    // 尝试解析为 JSON
                    try {
                        // 先尝试直接解析
                        return JSON.parse(langObjStr);
                    } catch (e) {
                        // 如果失败，尝试使用 eval（仅构建时）
                        try {
                            const vm = require('vm');
                            const sandbox = {};
                            vm.createContext(sandbox);
                            vm.runInContext(`const obj = ${langObjStr};`, sandbox);
                            return sandbox.obj;
                        } catch (e2) {
                            console.error(`Failed to parse ${langKey}:`, e2.message);
                            return null;
                        }
                    }
                }
                braceCount--;
            }
        }
    }
    
    return null;
}

// 定义要提取的语言列表
const languages = ['en', 'ru', 'pt', 'es', 'fr', 'ko', 'de', 'ja'];

// 使用更简单的方法：直接读取文件，找到每个语言对象的位置
console.log('Extracting language files...\n');

languages.forEach(lang => {
    // 查找语言对象的开始位置
    const langPattern = new RegExp(`\\s+${lang}:\\s*\\{`);
    const match = cleanedI18n.match(langPattern);
    
    if (!match) {
        console.warn(`⚠ Language ${lang} not found`);
        return;
    }
    
    const startIndex = cleanedI18n.indexOf(match[0]);
    if (startIndex === -1) {
        console.warn(`⚠ Could not find start position for ${lang}`);
        return;
    }
    
    // 从开始位置查找匹配的大括号
    let braceCount = 0;
    let inString = false;
    let stringChar = null;
    let escapeNext = false;
    
    // 找到第一个 { 的位置
    let objStart = startIndex + match[0].length - 1;
    for (let i = objStart; i < cleanedI18n.length; i++) {
        if (cleanedI18n[i] === '{') {
            objStart = i;
            break;
        }
    }
    
    for (let i = objStart; i < cleanedI18n.length; i++) {
        const char = cleanedI18n[i];
        
        if (escapeNext) {
            escapeNext = false;
            continue;
        }
        
        if (char === '\\') {
            escapeNext = true;
            continue;
        }
        
        if (!inString && (char === '"' || char === "'" || char === '`')) {
            inString = true;
            stringChar = char;
            continue;
        }
        
        if (inString && char === stringChar) {
            inString = false;
            stringChar = null;
            continue;
        }
        
        if (!inString) {
            if (char === '{') {
                braceCount++;
            } else if (char === '}') {
                braceCount--;
                if (braceCount === 0) {
                    // 找到匹配的结束大括号
                    const endIndex = i + 1;
                    const langObjStr = cleanedI18n.substring(objStart, endIndex);
                    
                    // 尝试解析
                    try {
                        // 使用 eval 解析（仅构建时，安全）
                        const vm = require('vm');
                        const sandbox = {};
                        vm.createContext(sandbox);
                        vm.runInContext(`const obj = ${langObjStr};`, sandbox);
                        
                        // 保存为 JSON
                        const jsonPath = path.join(localesDir, `${lang}.json`);
                        fs.writeFileSync(jsonPath, JSON.stringify(sandbox.obj, null, 2), 'utf-8');
                        console.log(`✓ Extracted ${lang}.json`);
                    } catch (error) {
                        console.error(`✗ Failed to extract ${lang}:`, error.message);
                    }
                    break;
                }
            }
        }
    }
});

console.log('\n✅ Language extraction completed!');
console.log(`📁 Files saved to: ${localesDir}`);
